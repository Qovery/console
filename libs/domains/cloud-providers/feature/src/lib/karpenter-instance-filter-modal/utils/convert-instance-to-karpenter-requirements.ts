import {
  type ClusterInstanceTypeResponseListResultsInner,
  type KarpenterNodePoolRequirement,
} from 'qovery-typescript-axios'

export function convertToKarpenterRequirements(
  instance: ClusterInstanceTypeResponseListResultsInner[]
): KarpenterNodePoolRequirement[] {
  if (!instance || instance.length === 0) {
    return []
  }

  const requirements: KarpenterNodePoolRequirement[] = []

  const uniqueValues = {
    architectures: new Set<string>(),
    families: new Set<string>(),
    sizes: new Set<string>(),
  }

  instance.forEach((instance) => {
    if (instance.architecture) {
      uniqueValues.architectures.add(instance.architecture)
    }

    if (instance.attributes?.instance_family) {
      uniqueValues.families.add(instance.attributes.instance_family)
    }
    if (instance.attributes?.instance_size) {
      uniqueValues.sizes.add(instance.attributes.instance_size)
    }
  })

  if (uniqueValues.sizes.size > 0) {
    requirements.push({
      key: 'InstanceSize',
      operator: 'In',
      values: Array.from(uniqueValues.sizes),
    })
  }

  if (uniqueValues.architectures.size > 0) {
    requirements.push({
      key: 'Arch',
      operator: 'In',
      values: Array.from(uniqueValues.architectures),
    })
  }

  if (uniqueValues.families.size > 0) {
    requirements.push({
      key: 'InstanceFamily',
      operator: 'In',
      values: Array.from(uniqueValues.families),
    })
  }

  return requirements
}
