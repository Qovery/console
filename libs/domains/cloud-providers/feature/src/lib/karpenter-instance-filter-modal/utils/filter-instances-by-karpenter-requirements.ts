import { type ClusterInstanceTypeResponseListResultsInner } from 'qovery-typescript-axios'
import { type KarpenterData } from '@qovery/shared/interfaces'

export function filterInstancesByKarpenterRequirements(
  allInstances: ClusterInstanceTypeResponseListResultsInner[],
  karpenterData: Omit<KarpenterData, 'disk_size_in_gib' | 'enabled' | 'spot_enabled'>
): ClusterInstanceTypeResponseListResultsInner[] {
  const karpenterRequirements = karpenterData.qovery_node_pools?.requirements || []

  return allInstances.filter((instance) => {
    const instanceSize = instance.attributes?.instance_size || ''
    const InstanceFamily = instance.attributes?.instance_family || ''
    const instanceArchitecture = instance.architecture || ''

    return karpenterRequirements.every((requirement) => {
      if (requirement.key === 'InstanceSize') {
        return requirement.values.includes(instanceSize)
      }

      if (requirement.key === 'Arch') {
        // Fix inconsistency between uppercase and lowercase with architecture during the migration
        return requirement.values.map((v) => v.toUpperCase()).includes(instanceArchitecture)
      }

      if (requirement.key === 'InstanceFamily') {
        return requirement.values.includes(InstanceFamily)
      }

      return false
    })
  })
}

export default filterInstancesByKarpenterRequirements
