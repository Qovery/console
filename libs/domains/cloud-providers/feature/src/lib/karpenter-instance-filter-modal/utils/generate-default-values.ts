import { type ClusterInstanceTypeResponseListResultsInner } from 'qovery-typescript-axios'
import { type KarpenterInstanceFormProps } from '../karpenter-instance-filter-modal'

export function generateDefaultValues(
  instances: ClusterInstanceTypeResponseListResultsInner[]
): Omit<KarpenterInstanceFormProps, 'default_service_architecture'> {
  // Get unique architectures
  const architectures = [...new Set(instances.map((i) => i.architecture))]

  // Get unique sizes without sorting
  const sizes = [...new Set(instances.map((i) => i.attributes?.instance_size ?? ''))].filter((i) => i !== '')

  // Group instance families by raw instance_category
  const categories: Record<string, string[]> = {}
  instances.forEach((instance) => {
    const family = instance.attributes?.instance_family || ''
    const category = instance.attributes?.instance_category || ''

    if (!categories[category]) {
      categories[category] = []
    }

    if (!categories[category].includes(family)) {
      categories[category].push(family)
    }
  })

  return {
    ARM64: architectures.includes('ARM64'),
    AMD64: architectures.includes('AMD64'),
    categories,
    sizes,
  }
}

export default generateDefaultValues
