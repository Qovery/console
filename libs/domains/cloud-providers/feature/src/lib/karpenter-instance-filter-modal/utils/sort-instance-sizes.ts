type BasicSize = 'small' | 'medium' | 'large' | 'xlarge'
type SizePriority = Record<BasicSize, number>

// Define type for the sort key tuple
type SortKey = [priority: number, multiplier: number]

export function sortInstanceSizes(sizes: string[]): string[] {
  const sizePriority: SizePriority = {
    small: 1,
    medium: 2,
    large: 3,
    xlarge: 4,
  }

  function getSortKey(size: string): SortKey {
    if (isBasicSize(size)) {
      return [sizePriority[size], 0]
    }

    // Handle xlarge variants
    if (size.includes('xlarge')) {
      // Extract number before 'xlarge'
      const number = size.replace('xlarge', '')
      // Convert empty string to 1 (for just 'xlarge')
      const multiplier = number ? parseInt(number) : 1
      return [5, multiplier]
    }

    return [0, 0]
  }

  function isBasicSize(size: string): size is BasicSize {
    return size in sizePriority
  }

  return sizes.sort((a, b) => {
    const [priorityA, multiplierA] = getSortKey(a)
    const [priorityB, multiplierB] = getSortKey(b)

    if (priorityA !== priorityB) {
      return priorityA - priorityB
    }
    return multiplierA - multiplierB
  })
}

export default sortInstanceSizes
