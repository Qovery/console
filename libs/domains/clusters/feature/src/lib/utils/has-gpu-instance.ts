import { type Cluster } from 'qovery-typescript-axios'

export const hasGpuInstance = (cluster?: Cluster) => {
  const clusterFeatureKarpenter = cluster?.features?.find((feature) => feature.id === 'KARPENTER')
  if (!clusterFeatureKarpenter) return false
  return Boolean(
    typeof clusterFeatureKarpenter?.value_object?.value === 'object' &&
      'qovery_node_pools' in clusterFeatureKarpenter.value_object.value &&
      clusterFeatureKarpenter.value_object.value.qovery_node_pools.gpu_override
  )
}
