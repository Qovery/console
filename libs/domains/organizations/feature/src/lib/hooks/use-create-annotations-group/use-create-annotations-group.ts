import { useMutation, useQueryClient } from '@tanstack/react-query'
import { mutations } from '@qovery/domains/organizations/data-access'
import { queries } from '@qovery/state/util-queries'

export function useCreateAnnotationsGroup() {
  const queryClient = useQueryClient()

  return useMutation(mutations.createAnnotationsGroup, {
    onSuccess(_, { organizationId }) {
      queryClient.invalidateQueries({
        queryKey: queries.organizations.annotationsGroups({ organizationId }).queryKey,
      })
    },
    meta: {
      notifyOnSuccess: {
        title: 'Your annotations group has been created',
      },
      notifyOnError: true,
    },
  })
}

export default useCreateAnnotationsGroup
