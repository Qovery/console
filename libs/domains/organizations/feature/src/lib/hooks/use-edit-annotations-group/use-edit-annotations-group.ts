import { useMutation, useQueryClient } from '@tanstack/react-query'
import { mutations } from '@qovery/domains/organizations/data-access'
import { queries } from '@qovery/state/util-queries'

export function useEditAnnotationsGroup() {
  const queryClient = useQueryClient()

  return useMutation(mutations.editAnnotationsGroup, {
    onSuccess(_, { organizationId }) {
      queryClient.invalidateQueries({
        queryKey: queries.organizations.annotationsGroups({ organizationId }).queryKey,
      })
    },
    meta: {
      notifyOnSuccess: {
        title:
          'Your annotations group has been edited. You need to redeploy your services using it for your changes to be applied.',
      },
      notifyOnError: true,
    },
  })
}

export default useEditAnnotationsGroup
