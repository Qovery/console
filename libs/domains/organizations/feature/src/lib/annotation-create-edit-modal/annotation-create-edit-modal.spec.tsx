import { wrapWithReactHookForm } from '__tests__/utils/wrap-with-react-hook-form'
import { OrganizationAnnotationsGroupScopeEnum } from 'qovery-typescript-axios'
import { renderWithProviders, screen } from '@qovery/shared/util-tests'
import * as useCreateAnnotationsGroup from '../hooks/use-create-annotations-group/use-create-annotations-group'
import * as useEditAnnotationsGroup from '../hooks/use-edit-annotations-group/use-edit-annotations-group'
import AnnotationCreateEditModal, {
  type AnnotationCreateEditModalProps,
  convertScopeEnumToObject,
  convertScopeObjectToEnum,
} from './annotation-create-edit-modal'

const useCreateAnnotationsGroupMockSpy = jest.spyOn(useCreateAnnotationsGroup, 'useCreateAnnotationsGroup') as jest.Mock
const useEditAnnotationsGroupMockSpy = jest.spyOn(useEditAnnotationsGroup, 'useEditAnnotationsGroup') as jest.Mock

const props: AnnotationCreateEditModalProps = {
  organizationId: '0000-0000-0000',
  onClose: jest.fn(),
}

describe('AnnotationCreateEditModal', () => {
  beforeEach(() => {
    useCreateAnnotationsGroupMockSpy.mockReturnValue({
      mutateAsync: jest.fn().mockResolvedValue({ id: '000' }),
    })
    useEditAnnotationsGroupMockSpy.mockReturnValue({
      mutateAsync: jest.fn(),
    })
  })

  it('should match with snatpshot', () => {
    const { baseElement } = renderWithProviders(wrapWithReactHookForm(<AnnotationCreateEditModal {...props} />))
    expect(baseElement).toMatchSnapshot()
  })

  it('should correctly convert array of enums to object', () => {
    const input = [OrganizationAnnotationsGroupScopeEnum.HPA, OrganizationAnnotationsGroupScopeEnum.CRON_JOBS]
    const expected = { HPA: true, CRON_JOBS: true }
    expect(convertScopeEnumToObject(input)).toMatchObject(expected)
  })

  it('should correctly convert object to array of enums', () => {
    const input = { HPA: true, CRON_JOBS: true }
    const expected = [OrganizationAnnotationsGroupScopeEnum.HPA, OrganizationAnnotationsGroupScopeEnum.CRON_JOBS]
    expect(convertScopeObjectToEnum(input)).toEqual(expected)
  })

  it('should submit the form to create a annotation group', async () => {
    const { userEvent } = renderWithProviders(wrapWithReactHookForm(<AnnotationCreateEditModal {...props} />))

    const inputName = screen.getByLabelText('Group name')
    await userEvent.type(inputName, 'my-name')

    const inputKey = screen.getByTestId('annotations.0.key')
    await userEvent.type(inputKey, 'key')

    const inputValue = screen.getByTestId('annotations.0.value')
    await userEvent.type(inputValue, 'value')

    const checkboxScope = screen.getByText('HPA')
    await userEvent.click(checkboxScope)

    const btn = screen.getByRole('button', { name: 'Create' })
    expect(btn).toBeEnabled()

    await userEvent.click(btn)

    expect(useCreateAnnotationsGroupMockSpy().mutateAsync).toHaveBeenCalledWith({
      organizationId: '0000-0000-0000',
      annotationsGroupRequest: {
        name: 'my-name',
        annotations: [
          {
            key: 'key',
            value: 'value',
          },
        ],
        scopes: ['HPA'],
      },
    })

    expect(props.onClose).toHaveBeenCalled()
  })

  it('should submit the form to edit a annotations group', async () => {
    const { userEvent } = renderWithProviders(
      wrapWithReactHookForm(
        <AnnotationCreateEditModal
          {...props}
          isEdit
          annotationsGroup={{
            id: '1111-1111-1111',
            created_at: '',
            updated_at: '',
            name: 'name',
            annotations: [
              {
                key: 'key',
                value: 'value',
              },
            ],
            scopes: ['HPA', 'CRON_JOBS'],
          }}
        />
      )
    )

    const inputName = screen.getByLabelText('Group name')
    await userEvent.clear(inputName)
    await userEvent.type(inputName, 'my-name')

    const btn = screen.getByRole('button', { name: 'Confirm' })
    expect(btn).toBeEnabled()

    await userEvent.click(btn)

    expect(useEditAnnotationsGroupMockSpy().mutateAsync).toHaveBeenCalledWith({
      organizationId: '0000-0000-0000',
      annotationsGroupId: '1111-1111-1111',
      annotationsGroupRequest: {
        name: 'my-name',
        annotations: [
          {
            key: 'key',
            value: 'value',
          },
        ],
        scopes: ['HPA', 'CRON_JOBS'],
      },
    })

    expect(props.onClose).toHaveBeenCalled()
  })
})
