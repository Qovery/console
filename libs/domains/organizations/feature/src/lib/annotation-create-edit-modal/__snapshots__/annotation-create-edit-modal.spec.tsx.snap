// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnnotationCreateEditModal should match with snatpshot 1`] = `
<body>
  <div>
    <div
      class="p-5"
    >
      <h2
        class="h4 max-w-sm truncate text-neutral-400 dark:text-neutral-50"
      >
        Create annotation group
      </h2>
      <p
        class="mt-2 text-sm text-neutral-350 dark:text-neutral-50"
      >
        You will have the possibility to modify the parameters once created.
      </p>
      <form
        class="mt-6"
      >
        <div
          class="mb-5"
          data-testid="input-text-wrapper"
        >
          <div
            class="relative"
          >
            <div
              aria-label="input-container"
              class="input     "
            >
              <div
                class=""
              >
                <label
                  class="translate-y-2 text-sm"
                  for="Group name"
                >
                  Group name
                </label>
                <input
                  class="input__value "
                  data-testid="input-text"
                  id="Group name"
                  name="name"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <span
          class="mb-2 block text-sm text-neutral-400"
        >
          Add annotation key value
        </span>
        <div
          class="rounded border border-neutral-250 bg-neutral-100 px-4 py-3"
        >
          <ul>
            <li
              class="mb-3 grid grid-cols-[6fr_6fr_1fr] gap-x-2"
            >
              <span
                class="text-sm font-medium text-neutral-350"
              >
                Annotation keys
              </span>
              <span
                class="text-sm font-medium text-neutral-350"
              >
                Value
              </span>
              <span />
            </li>
            <li
              class="mb-3 last:mb-0"
            >
              <div
                class="grid grid-cols-[6fr_6fr_1fr] items-center gap-x-2"
              >
                <div
                  class=""
                  data-testid="input-small-wrapper"
                >
                  <div
                    class="input input--small flex-grow"
                    data-testid="input"
                  >
                    <label
                      class="hidden"
                    />
                    <input
                      class="absolute left-0 top-0 h-full w-full rounded px-2 text-sm text-neutral-400 placeholder:text-neutral-350"
                      data-testid="annotations.0.key"
                      name="annotations.0.key"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
                <div
                  class=""
                  data-testid="input-small-wrapper"
                >
                  <div
                    class="input input--small flex-grow"
                    data-testid="input"
                  >
                    <label
                      class="hidden"
                    />
                    <input
                      class="absolute left-0 top-0 h-full w-full rounded px-2 text-sm text-neutral-400 placeholder:text-neutral-350"
                      data-testid="annotations.0.value"
                      name="annotations.0.value"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
                <button
                  class="inline-flex items-center shrink-0 font-medium transition focus-visible:[&:not(:active)]:outline-2 outline-0 hover:[&:not(:focus-visible):not(:disabled)]:shadow-sm active:[&:not(:focus-visible)]:shadow-none outline-brand-500 text-sm h-9 px-3 rounded border border-transparent text-neutral-350 hover:[&:not(:active):not(:focus-visible)]:border-neutral-300 hover:[&:not(:disabled)]:text-neutral-400 dark:hover:[&:not(:disabled)]:text-neutral-250 focus-visible:text-neutral-400 active:bg-neutral-150 dark:active:bg-transparent active:text-neutral-400 disabled:text-neutral-300 disabled:bg-neutral-150 disabled:border-none"
                  type="button"
                >
                  <i
                    aria-hidden="true"
                    class="fa-solid fa-trash text-neutral-400"
                  />
                </button>
              </div>
            </li>
          </ul>
          <button
            class="inline-flex items-center shrink-0 font-medium transition focus-visible:[&:not(:active)]:outline-2 outline-0 shadow-sm disabled:shadow-none active:shadow-none outline-neutral-500 text-xs h-7 px-2 rounded bg-brand-500 active:bg-brand-600 hover:bg-brand-400 text-white disabled:text-brand-300 disabled:bg-brand-100 mt-3"
            type="button"
          >
            Add annotation
            <i
              aria-hidden="true"
              class="fa-solid fa-plus ml-2"
            />
          </button>
        </div>
        <span
          class="mb-2 mt-4 block text-sm text-neutral-400"
        >
          Select scope (Kubernetes objects)
        </span>
        <div
          class="h-[194px] overflow-y-auto rounded border border-neutral-250 bg-neutral-100 px-4 py-3"
        >
          <div
            class="mb-2 flex items-center last:mb-0"
          >
            <button
              aria-checked="false"
              class="group transition bg-neutral-100 dark:bg-neutral-500 border border-neutral-300 dark:border-neutral-350 rounded-sm hover:border-2 data-[state=checked]:border data-[state=indeterminate]:border-2 disabled:border disabled:border-neutral-250 hover:border-brand-500 dark:hover:border-brand-500 data-[state=checked]:bg-brand-500 dark:data-[state=checked]:bg-brand-500 data-[state=checked]:border-brand-500 dark:data-[state=checked]:border-brand-500 data-[state=indeterminate]:border-brand-500 disabled:data-[state=checked]:border-brand-300 disabled:data-[state=checked]:bg-brand-300 disabled:data-[state=indeterminate]:border-brand-300 mr-3 h-4 w-4"
              data-state="unchecked"
              id="CRON_JOBS"
              role="checkbox"
              type="button"
              value="on"
            />
            <input
              aria-hidden="true"
              name="scopes.CRON_JOBS"
              style="transform: translateX(-100%); position: absolute; pointer-events: none; opacity: 0; margin: 0px; width: 0px; height: 0px;"
              tabindex="-1"
              type="checkbox"
              value="on"
            />
            <label
              class="text-sm font-medium text-neutral-400"
              for="CRON_JOBS"
            >
              CRON_JOBS
            </label>
          </div>
          <div
            class="mb-2 flex items-center last:mb-0"
          >
            <button
              aria-checked="false"
              class="group transition bg-neutral-100 dark:bg-neutral-500 border border-neutral-300 dark:border-neutral-350 rounded-sm hover:border-2 data-[state=checked]:border data-[state=indeterminate]:border-2 disabled:border disabled:border-neutral-250 hover:border-brand-500 dark:hover:border-brand-500 data-[state=checked]:bg-brand-500 dark:data-[state=checked]:bg-brand-500 data-[state=checked]:border-brand-500 dark:data-[state=checked]:border-brand-500 data-[state=indeterminate]:border-brand-500 disabled:data-[state=checked]:border-brand-300 disabled:data-[state=checked]:bg-brand-300 disabled:data-[state=indeterminate]:border-brand-300 mr-3 h-4 w-4"
              data-state="unchecked"
              id="DEPLOYMENTS"
              role="checkbox"
              type="button"
              value="on"
            />
            <input
              aria-hidden="true"
              name="scopes.DEPLOYMENTS"
              style="transform: translateX(-100%); position: absolute; pointer-events: none; opacity: 0; margin: 0px; width: 0px; height: 0px;"
              tabindex="-1"
              type="checkbox"
              value="on"
            />
            <label
              class="text-sm font-medium text-neutral-400"
              for="DEPLOYMENTS"
            >
              DEPLOYMENTS
            </label>
          </div>
          <div
            class="mb-2 flex items-center last:mb-0"
          >
            <button
              aria-checked="false"
              class="group transition bg-neutral-100 dark:bg-neutral-500 border border-neutral-300 dark:border-neutral-350 rounded-sm hover:border-2 data-[state=checked]:border data-[state=indeterminate]:border-2 disabled:border disabled:border-neutral-250 hover:border-brand-500 dark:hover:border-brand-500 data-[state=checked]:bg-brand-500 dark:data-[state=checked]:bg-brand-500 data-[state=checked]:border-brand-500 dark:data-[state=checked]:border-brand-500 data-[state=indeterminate]:border-brand-500 disabled:data-[state=checked]:border-brand-300 disabled:data-[state=checked]:bg-brand-300 disabled:data-[state=indeterminate]:border-brand-300 mr-3 h-4 w-4"
              data-state="unchecked"
              id="HPA"
              role="checkbox"
              type="button"
              value="on"
            />
            <input
              aria-hidden="true"
              name="scopes.HPA"
              style="transform: translateX(-100%); position: absolute; pointer-events: none; opacity: 0; margin: 0px; width: 0px; height: 0px;"
              tabindex="-1"
              type="checkbox"
              value="on"
            />
            <label
              class="text-sm font-medium text-neutral-400"
              for="HPA"
            >
              HPA
            </label>
          </div>
          <div
            class="mb-2 flex items-center last:mb-0"
          >
            <button
              aria-checked="false"
              class="group transition bg-neutral-100 dark:bg-neutral-500 border border-neutral-300 dark:border-neutral-350 rounded-sm hover:border-2 data-[state=checked]:border data-[state=indeterminate]:border-2 disabled:border disabled:border-neutral-250 hover:border-brand-500 dark:hover:border-brand-500 data-[state=checked]:bg-brand-500 dark:data-[state=checked]:bg-brand-500 data-[state=checked]:border-brand-500 dark:data-[state=checked]:border-brand-500 data-[state=indeterminate]:border-brand-500 disabled:data-[state=checked]:border-brand-300 disabled:data-[state=checked]:bg-brand-300 disabled:data-[state=indeterminate]:border-brand-300 mr-3 h-4 w-4"
              data-state="unchecked"
              id="INGRESS"
              role="checkbox"
              type="button"
              value="on"
            />
            <input
              aria-hidden="true"
              name="scopes.INGRESS"
              style="transform: translateX(-100%); position: absolute; pointer-events: none; opacity: 0; margin: 0px; width: 0px; height: 0px;"
              tabindex="-1"
              type="checkbox"
              value="on"
            />
            <label
              class="text-sm font-medium text-neutral-400"
              for="INGRESS"
            >
              INGRESS
            </label>
          </div>
          <div
            class="mb-2 flex items-center last:mb-0"
          >
            <button
              aria-checked="false"
              class="group transition bg-neutral-100 dark:bg-neutral-500 border border-neutral-300 dark:border-neutral-350 rounded-sm hover:border-2 data-[state=checked]:border data-[state=indeterminate]:border-2 disabled:border disabled:border-neutral-250 hover:border-brand-500 dark:hover:border-brand-500 data-[state=checked]:bg-brand-500 dark:data-[state=checked]:bg-brand-500 data-[state=checked]:border-brand-500 dark:data-[state=checked]:border-brand-500 data-[state=indeterminate]:border-brand-500 disabled:data-[state=checked]:border-brand-300 disabled:data-[state=checked]:bg-brand-300 disabled:data-[state=indeterminate]:border-brand-300 mr-3 h-4 w-4"
              data-state="unchecked"
              id="JOBS"
              role="checkbox"
              type="button"
              value="on"
            />
            <input
              aria-hidden="true"
              name="scopes.JOBS"
              style="transform: translateX(-100%); position: absolute; pointer-events: none; opacity: 0; margin: 0px; width: 0px; height: 0px;"
              tabindex="-1"
              type="checkbox"
              value="on"
            />
            <label
              class="text-sm font-medium text-neutral-400"
              for="JOBS"
            >
              JOBS
            </label>
          </div>
          <div
            class="mb-2 flex items-center last:mb-0"
          >
            <button
              aria-checked="false"
              class="group transition bg-neutral-100 dark:bg-neutral-500 border border-neutral-300 dark:border-neutral-350 rounded-sm hover:border-2 data-[state=checked]:border data-[state=indeterminate]:border-2 disabled:border disabled:border-neutral-250 hover:border-brand-500 dark:hover:border-brand-500 data-[state=checked]:bg-brand-500 dark:data-[state=checked]:bg-brand-500 data-[state=checked]:border-brand-500 dark:data-[state=checked]:border-brand-500 data-[state=indeterminate]:border-brand-500 disabled:data-[state=checked]:border-brand-300 disabled:data-[state=checked]:bg-brand-300 disabled:data-[state=indeterminate]:border-brand-300 mr-3 h-4 w-4"
              data-state="unchecked"
              id="PODS"
              role="checkbox"
              type="button"
              value="on"
            />
            <input
              aria-hidden="true"
              name="scopes.PODS"
              style="transform: translateX(-100%); position: absolute; pointer-events: none; opacity: 0; margin: 0px; width: 0px; height: 0px;"
              tabindex="-1"
              type="checkbox"
              value="on"
            />
            <label
              class="text-sm font-medium text-neutral-400"
              for="PODS"
            >
              PODS
            </label>
          </div>
          <div
            class="mb-2 flex items-center last:mb-0"
          >
            <button
              aria-checked="false"
              class="group transition bg-neutral-100 dark:bg-neutral-500 border border-neutral-300 dark:border-neutral-350 rounded-sm hover:border-2 data-[state=checked]:border data-[state=indeterminate]:border-2 disabled:border disabled:border-neutral-250 hover:border-brand-500 dark:hover:border-brand-500 data-[state=checked]:bg-brand-500 dark:data-[state=checked]:bg-brand-500 data-[state=checked]:border-brand-500 dark:data-[state=checked]:border-brand-500 data-[state=indeterminate]:border-brand-500 disabled:data-[state=checked]:border-brand-300 disabled:data-[state=checked]:bg-brand-300 disabled:data-[state=indeterminate]:border-brand-300 mr-3 h-4 w-4"
              data-state="unchecked"
              id="SECRETS"
              role="checkbox"
              type="button"
              value="on"
            />
            <input
              aria-hidden="true"
              name="scopes.SECRETS"
              style="transform: translateX(-100%); position: absolute; pointer-events: none; opacity: 0; margin: 0px; width: 0px; height: 0px;"
              tabindex="-1"
              type="checkbox"
              value="on"
            />
            <label
              class="text-sm font-medium text-neutral-400"
              for="SECRETS"
            >
              SECRETS
            </label>
          </div>
          <div
            class="mb-2 flex items-center last:mb-0"
          >
            <button
              aria-checked="false"
              class="group transition bg-neutral-100 dark:bg-neutral-500 border border-neutral-300 dark:border-neutral-350 rounded-sm hover:border-2 data-[state=checked]:border data-[state=indeterminate]:border-2 disabled:border disabled:border-neutral-250 hover:border-brand-500 dark:hover:border-brand-500 data-[state=checked]:bg-brand-500 dark:data-[state=checked]:bg-brand-500 data-[state=checked]:border-brand-500 dark:data-[state=checked]:border-brand-500 data-[state=indeterminate]:border-brand-500 disabled:data-[state=checked]:border-brand-300 disabled:data-[state=checked]:bg-brand-300 disabled:data-[state=indeterminate]:border-brand-300 mr-3 h-4 w-4"
              data-state="unchecked"
              id="SERVICES"
              role="checkbox"
              type="button"
              value="on"
            />
            <input
              aria-hidden="true"
              name="scopes.SERVICES"
              style="transform: translateX(-100%); position: absolute; pointer-events: none; opacity: 0; margin: 0px; width: 0px; height: 0px;"
              tabindex="-1"
              type="checkbox"
              value="on"
            />
            <label
              class="text-sm font-medium text-neutral-400"
              for="SERVICES"
            >
              SERVICES
            </label>
          </div>
          <div
            class="mb-2 flex items-center last:mb-0"
          >
            <button
              aria-checked="false"
              class="group transition bg-neutral-100 dark:bg-neutral-500 border border-neutral-300 dark:border-neutral-350 rounded-sm hover:border-2 data-[state=checked]:border data-[state=indeterminate]:border-2 disabled:border disabled:border-neutral-250 hover:border-brand-500 dark:hover:border-brand-500 data-[state=checked]:bg-brand-500 dark:data-[state=checked]:bg-brand-500 data-[state=checked]:border-brand-500 dark:data-[state=checked]:border-brand-500 data-[state=indeterminate]:border-brand-500 disabled:data-[state=checked]:border-brand-300 disabled:data-[state=checked]:bg-brand-300 disabled:data-[state=indeterminate]:border-brand-300 mr-3 h-4 w-4"
              data-state="unchecked"
              id="STATEFUL_SETS"
              role="checkbox"
              type="button"
              value="on"
            />
            <input
              aria-hidden="true"
              name="scopes.STATEFUL_SETS"
              style="transform: translateX(-100%); position: absolute; pointer-events: none; opacity: 0; margin: 0px; width: 0px; height: 0px;"
              tabindex="-1"
              type="checkbox"
              value="on"
            />
            <label
              class="text-sm font-medium text-neutral-400"
              for="STATEFUL_SETS"
            >
              STATEFUL_SETS
            </label>
          </div>
        </div>
        <div
          class="mt-6 flex justify-end gap-3"
        >
          <button
            class="inline-flex items-center shrink-0 font-medium transition focus-visible:[&:not(:active)]:outline-2 outline-0 hover:[&:not(:focus-visible):not(:disabled)]:shadow-sm active:[&:not(:focus-visible)]:shadow-none outline-brand-500 text-sm h-11 px-5 rounded border border-transparent text-neutral-350 hover:[&:not(:active):not(:focus-visible)]:border-neutral-300 hover:[&:not(:disabled)]:text-neutral-400 dark:hover:[&:not(:disabled)]:text-neutral-250 focus-visible:text-neutral-400 active:bg-neutral-150 dark:active:bg-transparent active:text-neutral-400 disabled:text-neutral-300 disabled:bg-neutral-150 disabled:border-none"
            data-testid="cancel-button"
            type="button"
          >
            Cancel
          </button>
          <button
            class="inline-flex items-center shrink-0 font-medium transition focus-visible:[&:not(:active)]:outline-2 outline-0 shadow-sm disabled:shadow-none active:shadow-none outline-neutral-500 text-sm h-11 px-5 rounded bg-brand-500 active:bg-brand-600 hover:bg-brand-400 text-white disabled:text-brand-300 disabled:bg-brand-100"
            data-testid="submit-button"
            disabled=""
            type="submit"
          >
            Create
          </button>
        </div>
      </form>
    </div>
  </div>
</body>
`;
